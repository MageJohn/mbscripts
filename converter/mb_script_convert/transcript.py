from dataclasses import dataclass, fields
from typing import Literal, Optional, Any, TypeVar, Self, cast
import logging

logger = logging.getLogger(__name__)

type Tag = Literal["direction", "character", "parenthetical", "dialogue", "end"]
type Content = list[tuple[Tag, str]]


type HugoFrontmatter = dict[str, str | HugoFrontmatter]


@dataclass
class Metadata:
    episode_title: Optional[str] = None
    series: Optional[str] = None
    season: Optional[str] = None
    season_episode_number: Optional[str] = None
    date_published: Optional[str] = None
    cover_url: Optional[str] = None

    def merge(self, other: Self):
        for field in fields(self):
            if getattr(self, field.name) is None:
                setattr(self, field.name, getattr(other, field.name))

    def to_hugo_frontmatter(self) -> HugoFrontmatter:
        params = {}
        frontmatter: HugoFrontmatter = {"params": params}
        for field in fields(self):
            name, value = field.name, getattr(self, field.name)
            if value is None:
                continue
            match name:
                case "episode_title":
                    frontmatter["title"] = value
                case "date_published":
                    frontmatter["date"] = value
                case _:
                    params[name] = value
        return frontmatter

    @classmethod
    def from_hugo_frontmatter(cls, frontmatter: dict[str, Any]) -> Self:
        m = cls()

        params = _checked_get(frontmatter, "params", dict)
        if params is None:
            params = {}
        for field in fields(m):
            match field.name:
                case "episode_title":
                    m.episode_title = _checked_get(frontmatter, "title", str)
                case "date_published":
                    m.date_published = _checked_get(frontmatter, "date", str)
                case _:
                    setattr(m, field.name, _checked_get(params, field.name, cast(type, field.type)))
        return m


T = TypeVar("T")


def _checked_get(frontmatter: dict[str, Any], key: str, type_: type[T]) -> T | None:
    value = frontmatter.get(key)
    if isinstance(value, type_):
        return value
    elif value is None:
        return None
    else:
        logger.warning(
            f"Invalid type for field {key} when loading frontmatter. Expected {type_}, got {type(value)}."
        )


class Transcript:
    """
    Represents a transcript in a format agnostic way.
    Used as an intermediate step when converting.
    """

    content: Content
    metadata: Metadata

    def __init__(
        self, metadata: Metadata | None = None, content: Content | None = None
    ):
        self.metadata = Metadata() if metadata is None else metadata
        self.content = [] if content is None else content

    def add_content(self, tag: Tag, content: str):
        self.content.append((tag, content))
