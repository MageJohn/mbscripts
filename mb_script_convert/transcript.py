from typing import Literal

type Tag = Literal["direction", "character", "parenthetical", "dialogue", "end"]
type Content = list[tuple[Tag, str]]
type Metadata = dict[str, str | Metadata]


class Transcript:
    """
    Represents a transcript in a format agnostic way.
    Used as an intermediate step when converting.
    """

    content: Content
    metadata: Metadata

    def __init__(self, metadata: Metadata = {}, content: Content = []):
        self.metadata = metadata
        self.content = content

    def add_content(self, tag: Tag, content: str):
        self.content.append((tag, content))

    def set_metadata(self, key: str, value: str):
        self.metadata[key] = value
