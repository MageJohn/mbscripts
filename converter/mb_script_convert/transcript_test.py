import tomllib

from .transcript import Metadata

FRONTMATTER = """
title = "Chapter 1: The Transdimensional Haboob"
date = "2020-10-27T18:01:00"

[params]
series = "Midnight Burger"
season = "1"
season_episode_number = "1"
cover_url = "https://example.com/image.jpg"
"""


class TestMetadataFromFrontmatter:
    def test_from_toml(self, snapshot):
        frontmatter = tomllib.loads(FRONTMATTER)
        metadata = Metadata.from_hugo_frontmatter(frontmatter)

        assert metadata == snapshot

    def test_no_params(self, snapshot):
        frontmatter = {
            "title": "Chapter 42",
        }
        metadata = Metadata.from_hugo_frontmatter(frontmatter)
        assert metadata == Metadata(episode_title="Chapter 42")
