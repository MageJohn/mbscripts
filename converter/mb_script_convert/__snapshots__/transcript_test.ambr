# serializer version: 1
# name: TestMetadataFromFrontmatter.test_from_toml
  Metadata(episode_title='Chapter 1: The Transdimensional Haboob', series='Midnight Burger', season='1', season_episode_number='1', date_published='2020-10-27T18:01:00', cover_url='https://example.com/image.jpg')
# ---
